File: src/ts/ssh-tcp/messages/portForwardChannelOpenMessage.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { ChannelOpenMessage, SshDataReader, SshDataWriter } from '@microsoft/dev-tunnels-ssh';

export class PortForwardChannelOpenMessage extends ChannelOpenMessage {
	public host: string = '';
	public port: number = 0;
	public originatorIPAddress: string = '';
	public originatorPort: number = 0;

	protected onRead(reader: SshDataReader): void {
		super.onRead(reader);

		this.host = reader.readString('ascii');
		this.port = reader.readUInt32();
		this.originatorIPAddress = reader.readString('ascii');
		this.originatorPort = reader.readUInt32();
	}

	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);

		writer.writeString(this.validateField(this.host, 'host'), 'ascii');
		writer.writeUInt32(this.validateField(this.port, 'port'));
		writer.writeString(this.originatorIPAddress || '', 'ascii');
		writer.writeUInt32(this.originatorPort || 0);
	}

	public toString() {
		return `${super.toString()} (host=${this.host} port=${this.port})`;
	}
}

File: src/cs/Ssh.Tcp/Messages/PortForwardChannelOpenMessage.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Text;
using Microsoft.DevTunnels.Ssh.IO;

namespace Microsoft.DevTunnels.Ssh.Messages;

public class PortForwardChannelOpenMessage : ChannelOpenMessage
{
	public string Host { get; set; } = string.Empty;
	public uint Port { get; set; }
	public string OriginatorIPAddress { get; set; } = string.Empty;
	public uint OriginatorPort { get; set; }

	protected override void OnRead(ref SshDataReader reader)
	{
		base.OnRead(ref reader);

		Host = reader.ReadString(Encoding.ASCII);
		Port = reader.ReadUInt32();
		OriginatorIPAddress = reader.ReadString(Encoding.ASCII);
		OriginatorPort = reader.ReadUInt32();
	}

	protected override void OnWrite(ref SshDataWriter writer)
	{
		base.OnWrite(ref writer);

		writer.Write(Host ?? string.Empty, Encoding.ASCII);
		writer.Write(Port);
		writer.Write(OriginatorIPAddress ?? string.Empty, Encoding.ASCII);
		writer.Write(OriginatorPort);
	}

	public override string ToString() =>
		$"{GetType().Name} ChannelType: {ChannelType}, SenderChannel: {SenderChannel}, " +
		$"Host: {Host}, Port: {Port}, OriginatorIPAddress: {OriginatorIPAddress}, OriginatorPort: {OriginatorPort}";
}
